File: src/app/browser/main/tab/page/content/text/gemini.cpp
#include "gemini.hpp"
#include "gemini/reader.hpp"

using namespace app::browser::main::tab::page::content::text;

Gemini::Gemini(
    const Glib::RefPtr<Gio::SimpleAction> & ACTION__OPEN_LINK_VARIANT,
    const Glib::ustring & GEMTEXT,
    Glib::ustring & title,
    GUri * uri
) : Gtk::Viewport( // add scrolled window features to childs
    NULL,
    NULL
) {
    // Init widget
    set_scroll_to_focus(
        false
    );

    set_child(
        * Gtk::make_managed<gemini::Reader>(
            ACTION__OPEN_LINK_VARIANT,
            GEMTEXT,
            title,
            uri
        )
    );
}

File: src/browser/main/tab/page/content/text/gemini.rs
mod reader;

use reader::Reader;

use gtk::{
    gio::SimpleAction,
    glib::{GString, Uri},
    Viewport,
};

use std::sync::Arc;

pub struct Gemini {
    reader: Reader,
    widget: Viewport,
}

impl Gemini {
    // Construct
    pub fn new(gemtext: &str, base: &Uri, action_open: Arc<SimpleAction>) -> Self {
        // Init components
        let reader = Reader::new(gemtext, base, action_open);

        // Init widget
        let widget = Viewport::builder().scroll_to_focus(false).build();

        widget.set_child(Some(reader.widget()));

        // Result
        Self { reader, widget }
    }

    // Getters
    pub fn reader_title(&self) -> &Option<GString> {
        &self.reader.title()
    }

    pub fn widget(&self) -> &Viewport {
        &self.widget
    }
}
